### PR TITLE
More bug fixes for pow and expand

### LINKED ISSUE #370: CSYMPY_ASSERT() failure
I just discovered a major bug (the latest master, 2793ba5abedfd04c9573ad37da30f4aa02ff77f7):
```
In [1]: from csympy import *

In [2]: var("x y z")
Out[2]: (x, y, z)

In [3]: e = (x+y+sqrt(3)*z+1)**15

In [4]: f = e*(e+1)

In [5]: g = f.expand()
python2.7: /home/ondrej/repos/csympy/src/mul.cpp:16: CSymPy::Mul::Mul(const CSymPy::RCP<const CSymPy::Number>&, CSymPy::map_basic_basic&&): Assertion `is_canonical(coef, dict_)' failed.

Abort caught. Printing stacktrace:

Traceback (most recent call last):
  File unknown, in _start()
  File "/build/buildd/eglibc-2.19/csu/libc-start.c", line 287, in __libc_start_main()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Modules/main.c", line 640, in Py_Main()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/pythonrun.c", line 955, in PyRun_SimpleFileExFlags()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/pythonrun.c", line 1377, in run_mod()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 673, in PyEval_EvalCode()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4129, in fast_function()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4346, in ext_do_call()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Objects/abstract.c", line 2529, in PyObject_Call()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Objects/funcobject.c", line 526, in function_call()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4129, in fast_function()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4129, in fast_function()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4129, in fast_function()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4129, in fast_function()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4129, in fast_function()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4129, in fast_function()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4730, in exec_statement()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 673, in PyEval_EvalCode()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 3265, in PyEval_EvalCodeEx()
  File "/home/ondrej/.hashdist/tmp/python-ctzx2dm36c6d/Python/ceval.c", line 4017, in call_function()
  File "/home/ondrej/repos/csympy/csympy/lib/csympy_wrapper.cpp", line 5803, in __pyx_pw_6csympy_3lib_14csympy_wrapper_5Basic_27expand()
    __pyx_r = __pyx_pf_6csympy_3lib_14csympy_wrapper_5Basic_26expand(((struct __pyx_obj_6csympy_3lib_14csympy_wrapper_Basic *)__pyx_v_self));
  File "/home/ondrej/repos/csympy/csympy/lib/csympy_wrapper.cpp", line 5839, in __pyx_pf_6csympy_3lib_14csympy_wrapper_5Basic_26expand()
    __pyx_t_1 = CSymPy::expand(((CSymPy::RCP<CSymPy::Basic const > &)__pyx_v_self->thisptr));
  File "/home/ondrej/repos/csympy/src/basic.cpp", line 40, in CSymPy::expand(CSymPy::RCP<CSymPy::Basic const> const&)
    if (is_a<Mul>(*self)) return mul_expand(rcp_static_cast<const Mul>(self));
  File "/home/ondrej/repos/csympy/src/mul.cpp", line 493, in CSymPy::mul_expand(CSymPy::RCP<CSymPy::Mul const> const&)
    a = expand(a);
  File "/home/ondrej/repos/csympy/src/basic.cpp", line 41, in CSymPy::expand(CSymPy::RCP<CSymPy::Basic const> const&)
    if (is_a<Pow>(*self)) return pow_expand(rcp_static_cast<const Pow>(self));
  File "/home/ondrej/repos/csympy/src/pow.cpp", line 412, in CSymPy::pow_expand(CSymPy::RCP<CSymPy::Pow const> const&)
    RCP<const Basic> term = Mul::from_dict(overall_coeff, std::move(d));
  File "/home/ondrej/repos/csympy/src/mul.cpp", line 214, in CSymPy::Mul::from_dict(CSymPy::RCP<CSymPy::Number const> const&, std::map<CSymPy::RCP<CSymPy::Basic const>, CSymPy::RCP<CSymPy::Basic const>, CSymPy::RCPBasicKeyLess, std::allocator<std::pair<CSymPy::RCP<CSymPy::Basic const> const, CSymPy::RCP<CSymPy::Basic const> > > >&&)
    return rcp(new Mul(coef, std::move(d)));
  File "/home/ondrej/repos/csympy/src/mul.cpp", line 16, in CSymPy::Mul::Mul(CSymPy::RCP<CSymPy::Number const> const&, std::map<CSymPy::RCP<CSymPy::Basic const>, CSymPy::RCP<CSymPy::Basic const>, CSymPy::RCPBasicKeyLess, std::allocator<std::pair<CSymPy::RCP<CSymPy::Basic const> const, CSymPy::RCP<CSymPy::Basic const> > > >&&)
    CSYMPY_ASSERT(is_canonical(coef, dict_))
  File "/build/buildd/eglibc-2.19/assert/assert.c", line 101, in __GI___assert_fail()
  File "/build/buildd/eglibc-2.19/assert/assert.c", line 92, in __assert_fail_base()
  File "/build/buildd/eglibc-2.19/stdlib/abort.c", line 89, in __GI_abort()
  File "/build/buildd/eglibc-2.19/signal/../nptl/sysdeps/unix/sysv/linux/raise.c", line 56, in __GI_raise()
  File unknown, in killpg()
  File "/home/ondrej/repos/csympy/src/teuchos/Teuchos_stacktrace.cpp", line 449, in loc_abort_callback_print_stack()
    Teuchos::show_stacktrace();

Done.
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by isuruf**
Fix bug in mul_expand because of failure to initialize variable

Fixes #370.

### CODE REVIEW

**Comment by certik** [comment]
Thanks! I just verified that it fixes #370. I'll review it soon.

File: src/tests/basic/test_arit.cpp [hunk 1]
```diff
@@ -736,6 +736,29 @@ void test_expand2()
...
+    r1 = expand(r1);
+    std::cout << r1->__str__() << std::endl;
+
+    r1 = pow(add(y, mul(sqrt(i3), z)), i3);
+    r1 = expand(r1);
+    std::cout << r1->__str__() << std::endl;
+
+    r1 = pow(mul(sqrt(i3), mul(y, add(one, pow(i3, div(one, i3))))), i3);
+    r1 = expand(r1);
+    std::cout << r1->__str__() << std::endl;
```
Comment (certik) [inline src/tests/basic/test_arit.cpp]: Do these tests cover both of the bugs reported in #370?

File: src/mul.cpp [hunk 2]
```diff
@@ -251,9 +251,23 @@ void Mul::dict_add_term_new(const Ptr<RCP<const Number>> &coef, map_basic_basic
...
+            } else if (is_a<Rational>(*exp)) {
+                mpz_class q, r, num, den;
+                num = rcp_static_cast<const Rational>(exp)->i.get_num();
+                den = rcp_static_cast<const Rational>(exp)->i.get_den();
+                mpz_fdiv_qr(q.get_mpz_t(), r.get_mpz_t(), num.get_mpz_t(),
+                    den.get_mpz_t());
+
+                insert(d, t, Rational::from_mpq(mpq_class(r, den)));
+                imulnum(outArg(*coef), pownum(rcp_static_cast<const Number>(t),
+                    rcp_static_cast<const Number>(integer(q))));
```
Comment (certik) [inline src/mul.cpp]: What exactly is this code doing?

File: src/mul.cpp [hunk 3]
```diff
@@ -295,6 +309,18 @@ void Mul::dict_add_term_new(const Ptr<RCP<const Number>> &coef, map_basic_basic
...
+            mpz_class q, r, num, den;
+            num = rcp_static_cast<const Rational>(it->second)->i.get_num();
+            den = rcp_static_cast<const Rational>(it->second)->i.get_den();
+            if (num > den || num < 0) {
+                mpz_fdiv_qr(q.get_mpz_t(), r.get_mpz_t(), num.get_mpz_t(),
+                    den.get_mpz_t());
+
+                it->second = Rational::from_mpq(mpq_class(r, den));
+                imulnum(outArg(*coef), pownum(rcp_static_cast<const Number>(t),
+                    rcp_static_cast<const Number>(integer(q))));
```
Comment (certik) [inline src/mul.cpp]: Seems like a code repetition with the above code change. We can refactor in a new PR if there is a way.

**Comment by certik** [comment]
This is +1 to merge from me. There seems to be some code duplication, not just this PR, but also what's already in csympy, and maybe we can have a look in another PR if there is a way to clean this up.

But this PR looks good to me to merge and it fixes both of the problems and benchmarks don't seem to be negatively affected.

File: src/mul.cpp (hunk 2 above)
Comment (isuruf) [inline src/mul.cpp]: `exp = num / den = q + r / den`. We factor out `t ^ q` and put `t ^ (r / den)` into the dictionary.

Comment (certik) [inline src/mul.cpp]: Ah yes. We should put a comment into the code.

**Comment by certik** [comment]
+1 to merge if tests pass.

**Comment by certik** [comment]
Tests pass, so I am merging this, as it fixes an important bug.

**Pull request merged**